### PR TITLE
Fixes #15405 - Moved puppet UI to helpers

### PR DESCRIPTION
**Pull request by ShimShtein**
Created puppet-specific helpers as part of pulling puppet outside of core.

### CODE REVIEW

**Comment by ShimShtein** [comment]
Fixed tests and improved codeclimate a bit.

**Comment by ShimShtein** [comment]
Test failure doesn't seem related, let's retry...
[test]

File: app/helpers/puppet_related_helper.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    super
+
+    %w(hosts/new hosts/edit hosts/process_hostgroup).each do |action|
+      mgr = Pagelets::Manager.new action
+      mgr.add_pagelet :main_tabs,
+      :id => :puppet_klasses,
+      :name => _('Puppet Classes'),
+      :partial => 'hosts/puppet/puppet_classes_tab',
+      :priority => 100,
+      :onlyif => proc { SmartProxy.with_features("Puppet").count > 0 }
```
Comment (orrabin) [inline app/helpers/puppet_related_helper.rb]: I know you just moved this but does it make sense to add the puppet tab on this condition and the puppet ca/master fields according to the condition in `tax_scoped_smart_proxy`?
It looks like there are cases where we can get the ca/master without the tab and vice versa.
Maybe someone can give an example of why that makes sense and if not maybe create one condition for the three of them?

File: app/controllers/hosts_controller.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@ class HostsController < ApplicationController
   before_filter :validate_power_action, :only => :update_multiple_power_state
   before_filter :validate_multiple_puppet_proxy, :only => :update_multiple_puppet_proxy
   before_filter :validate_multiple_puppet_ca_proxy, :only => :update_multiple_puppet_ca_proxy
-  helper :hosts, :reports, :interfaces
+  helper :hosts, :hosts_and_hostgroups, :reports, :interfaces, :puppet_related
```
Comment (orrabin) [inline app/controllers/hosts_controller.rb]: If the order is important here what will happen with the other pagelets added later on? do they have to be added after the puppet one or just always keep them after hosts_and_hostgroups?

The reason for my question is that later on you use `super` to make sure `define_host_pagelets` isn't overridden but added to whatever was already loaded but if more pagelets are added the `super` can go to either puppet or  hosts_and_hostgroups. I'm assuming the order doesn't really matter as long as all of them are loaded but it is a little confusing to debug when super can lead to multiple paths.

File: app/helpers/puppet_related_helper.rb [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+             { :include_blank => blank_or_inherit_f(f, :puppet_ca_proxy),
+               :disable_button => can_override ? _(HostsAndHostgroupsHelper::INHERIT_TEXT) : nil,
+               :disable_button_enabled => override && !explicit_value?(:puppet_ca_proxy_id),
+               :user_set => user_set?(:puppet_ca_proxy_id)
+             },
+             { :label       => _("Puppet CA"),
+               :help_inline => _("Use this puppet server as a CA server") }
+  end
+
+  def puppet_master(f, can_override, override)
```
Comment (orrabin) [inline app/helpers/puppet_related_helper.rb]: This and `puppet_ca` look very similar, maybe there is a way to DRY it out?

**Comment by ShimShtein** [comment]
Found a new way to initialize pagelets, so removed initialization code.
DRY-ed the puppet and puppet ca fields

> I know you just moved this but does it make sense to add the puppet tab on this condition and the puppet ca/master fields according to the condition in tax_scoped_smart_proxy?
> It looks like there are cases where we can get the ca/master without the tab and vice versa.
> Maybe someone can give an example of why that makes sense and if not maybe create one condition for the three of them?

Is still a valid question. Anyone?

**Comment by orrabin** [comment]
@ShimShtein codeclimate is happy now and I like the new initialization way much better, I still need to test but my question can be a refactoring in a separate pr so it shouldn't hold this back.
I'm not getting the integration tests failures locally at all so it might be another intermittent failure.
Maybe [test] again is enough.

**Comment by orrabin** [comment]
@ShimShtein testing this I see the puppet classes tab and ca/master fields disappear and appear according to hostgroup. Steps to reproduce:
- go to new host
- you will see the fields and tab
- fill in name and organization
- the tab and fields disappeared
- fill in hostgroup
- now the tab and fields appeared again
- remove hostgroup and they are still there

**Comment by ShimShtein** [comment]
The failure is probably due to http://projects.theforeman.org/issues/15464.
We'll have to wait until this one is closed, otherwise we will chase the missing actions forever.

BTW, codeclimate is green because it has skipped analysis, not because the code got better...

**Comment by ShimShtein** [comment]
Rebased, fixed the pagelet subscription error.

**Comment by ShimShtein** [comment]
Reabsed again and re-tested again on my machine: no failures.

**Comment by ShimShtein** [comment]
Test results lost, retesting
[test]

**Comment by ShimShtein** [comment]
Rebased, let's see if the errors will come back.

**Comment by ShimShtein** [comment]
Tests fixed, added an option to isolate configuration registries.
Thanks @domcleal !

**Comment by ShimShtein** [comment]
Rebased

**Comment by ShimShtein** [comment]
[test foreman]

**Comment by ShimShtein** [comment]
Rebased. Removed special handling for puppet_ca and puppet_proxy fields, based on @timogoebel's generic proxies work.

**Comment by ShimShtein** [comment]
Removed duplicate definitions.

**Comment by ShimShtein** [comment]
@orrabin Could you review this PR again?

File: app/helpers/hosts_and_hostgroups_helper.rb [hunk 4]
```diff
@@ -1,32 +1,13 @@
 module HostsAndHostgroupsHelper
   include AncestryHelper
+  include AccessibleResourceHelper
```
Comment (tbrisker) [inline app/helpers/hosts_and_hostgroups_helper.rb]: Why are you including the helper again? The functions in ApplicationHelper are available on all pages.

File: app/helpers/puppet_related_helper.rb [hunk 5]
```diff
@@ -0,0 +1,114 @@
+module PuppetRelatedHelper
+  extend AccessibleResourceHelper
```
Comment (tbrisker) [inline app/helpers/puppet_related_helper.rb]: Why does this extend AccessibleResourceHelper?

File: app/helpers/puppet_related_helper.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+  # initialize puppet related pagelets
```
Comment (tbrisker) [inline app/helpers/puppet_related_helper.rb]: Shouldn't pagelet initialization be done in the initializers? Wouldn't this get executed every time the helper is loaded?

File: app/helpers/puppet_related_helper.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        render 'puppetclasses/class_selection', :obj => puppetclasses_receiver
+      else
+        alert(:class => 'alert-info', :header => _('Notice'),
+              :text => _('Please select an environment first'))
+      end
+    end
+  end
+
+  class << self
+    def accessible_smart_proxy(obj, feature, property)
```
Comment (tbrisker) [inline app/helpers/puppet_related_helper.rb]: Looks like there is some copy&paste error in this function? Or maybe a started refactoring that didn't finish? I would prefer if any refactoring was not done in this PR other then the minimum required for the extraction, as it is hard to notice these changes in such a large block of text that has been moved from one place to the other.

File: app/helpers/hosts_and_hostgroups_helper.rb [hunk 8]
```diff
@@ -62,81 +43,4 @@ def realm_field(f, can_override = false, override = false)
...
-    classes_ids = classes.reorder('').pluck('puppetclasses.id')
-    smart_vars = VariableLookupKey.reorder('').where(:puppetclass_id => classes_ids).uniq.pluck(:puppetclass_id)
-    class_vars = PuppetclassLookupKey.reorder('').joins(:environment_classes).where(:environment_classes => { :puppetclass_id => classes_ids }).uniq.pluck('environment_classes.puppetclass_id')
-    klasses    = (smart_vars + class_vars).uniq
-
-    classes.where(:id => klasses)
-  end
-
-  def puppetclasses_tab(puppetclasses_receiver)
-    return unless accessible_puppet_proxies(puppetclasses_receiver).present?
```
Comment (tbrisker) [inline app/helpers/hosts_and_hostgroups_helper.rb:131]: Why was this line removed?

File: test/helpers/pagelets_helper_test.rb [hunk 9]
```diff
@@ -1,17 +1,17 @@
 require 'test_helper'
+require 'statics_isolator_helper'
 
 class PageletsHelperTest < ActionView::TestCase
+  include Isolator
   include PageletsHelper
 
+  isolate_config Pagelets::Manager
```
Comment (tbrisker) [inline test/helpers/pagelets_helper_test.rb]: I'm not sure I understand the need for this whole module. Why not just define what you want to be defined for the specific text context in the setup block and clear it in the tear down block?

File: app/views/hosts/puppet/_main_tab_fields.html.erb [hunk 10]
```diff
@@ -0,0 +1 @@
+<%= host_puppet_environment_field form %>
```
Comment (tbrisker) [inline app/views/hosts/puppet/_main_tab_fields.html.erb:1]: Is there a way to have pagelets run a function instead of rendering a partial that calls a function?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Thanks, please see inline comments

File: app/helpers/hosts_and_hostgroups_helper.rb (hunk 8 above)
Comment (ShimShtein) [reply app/helpers/hosts_and_hostgroups_helper.rb:131]: it was moved to pagelet definition:

``` ruby
mgr.add_pagelet :main_tabs,
  :id => :puppet_klasses,
  :name => _('Puppet Classes'),
  :partial => 'hosts/puppet/puppet_classes_tab',
  :priority => 100,
  :onlyif => proc { |host| host.accessible_resource(SmartProxy, :association => :puppet_proxy).present? }
# The :onlyif statement ^^
```

File: app/helpers/puppet_related_helper.rb (hunk 6 above)
Comment (ShimShtein) [reply app/helpers/puppet_related_helper.rb]: Moved indeed

File: app/helpers/puppet_related_helper.rb (hunk 7 above)
Comment (ShimShtein) [reply app/helpers/puppet_related_helper.rb]: Removed completely

File: app/views/hosts/puppet/_main_tab_fields.html.erb (hunk 10 above)
Comment (ShimShtein) [reply app/views/hosts/puppet/_main_tab_fields.html.erb:1]: Not that I know of...
@xprazak2, IMHO it's a valid RFE.

File: test/helpers/pagelets_helper_test.rb (hunk 9 above)
Comment (ShimShtein) [reply test/helpers/pagelets_helper_test.rb]: The proble I was trying to solve, is preservation of pagelets' registry between tests - The original code had `Pagelets.Manager.clear` in the teardown, which removed pagelets registered outside the tests (my pagelets). The isolator remembers the state of the local registry, and restores it after the test is done.

**Comment by ShimShtein** [comment]
Made hound happy again.

Refactored resource accessibility methods into a concern for model. They should have been there in the first place, since they deal with data retrieval.

**Comment by ShimShtein** [comment]
@tbrisker, I think I have addressed all the things you have mentioned.

**Comment by tbrisker** [comment]
@ShimShtein The whole refactoring of the accessible helpers is unrelated to this issue, and in IMHO complicates the code instead of simplifying it. If you want to refactor it please do so in a separate PR, as I understand it this PR should just extract the puppet helpers from the host helpers, not refactor several other unrelated areas. Also rebase on current develop to get rid of the hound comment on single quoted strings.

File: test/helpers/pagelets_helper_test.rb (hunk 9 above)
Comment (tbrisker) [reply test/helpers/pagelets_helper_test.rb]: The problem is that instead of removing just the pagelets that were added this test clears all of them. The `add_pagelet` part should just be in a setup block, and if needed separated in several contexts.

Comment (tbrisker) [reply test/helpers/pagelets_helper_test.rb]: Or just add your pagelets in the tests that need them.

**Comment by ShimShtein** [comment]
Rebased.

> The whole refactoring of the accessible helpers is unrelated to this issue

Unfortunately I need it in the pagelet `:onlyif` statement (which is outside of helpers context).

**Comment by tbrisker** [comment]
Or leave the condition in the function instead of moving it to `:onlyif` for the pagelet?

**Comment by ShimShtein** [comment]
The `:onlyif` statement for tab pagelet will influence both tab label and tab content.

**Comment by ShimShtein** [comment]
If I use pagelets, I don't have control over tab label, as opposed to explicit definition as it was.

File: test/helpers/pagelets_helper_test.rb (hunk 9 above)
Comment (ShimShtein) [reply test/helpers/pagelets_helper_test.rb]: I am adding the pagelets in the initializer. Many tests are relying on the fact that puppet tabs are in place without extra setup, which sounds perfect to me (they should be there in prod). The problem here is that pagelet tests are clearing a global registry.
The same problem will happen when I'll move puppet models to facets - when the registry will be cleared in facet tests, it will loose facets registered in the initializers.

File: app/views/hosts/puppet/_main_tab_fields.html.erb (hunk 10 above)
Comment (xprazak2) [reply app/views/hosts/puppet/_main_tab_fields.html.erb:1]: I agree, it is a valid RFE.

File: test/helpers/pagelets_helper_test.rb (hunk 9 above)
Comment (tbrisker) [reply test/helpers/pagelets_helper_test.rb]: Since this is currently only used here, we can use a simpler solution to creating a generalized class. In the setup block add something like:
```ruby
@prev_state = Pagelets::Manager.instance_variable_get(:@pagelets).clone
```
then in the teardown, after the clear, 
```ruby
Pagelets::Manager.instance_variable_set :@pagelets, @prev_state
```
If you want to make sure the tests are isolated you can even add another clear to the setup block.

File: app/helpers/application_helper.rb [hunk 11]
```diff
@@ -1,6 +1,7 @@
 # Methods added to this helper will be available to all templates in the application
 module ApplicationHelper
   include HomeHelper
+  include AccessibleResourceHelper
```
Comment (tbrisker) [inline app/helpers/application_helper.rb]: This is doesn't exist any more?

File: app/models/concerns/resource_accessibility.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
+module ResourceAccessibility
```
Comment (tbrisker) [inline app/models/concerns/resource_accessibility.rb]: Instead of moving this logic to a concern, you can pass the view context to the pagelet `onlyif` Proc by passing `self` as an extra parameter when calling the proc from PageletHelper. This will allow you to use the existing helpers.

**Review by tbrisker — CHANGES_REQUESTED** [review]
See comments inline.

**Comment by ShimShtein** [comment]
Rebased and fixed as you suggested.

File: test/helpers/pagelets_helper_test.rb [hunk 13]
```diff
@@ -1,15 +1,18 @@
 require 'test_helper'
+require 'statics_isolator_helper'
```
Comment (orrabin) [inline test/helpers/pagelets_helper_test.rb]: leftover?

File: test/helpers/pagelets_helper_test.rb [hunk 14]
```diff
@@ -1,15 +1,18 @@
...
 
 class PageletsHelperTest < ActionView::TestCase
+  include Isolator
```
Comment (orrabin) [inline test/helpers/pagelets_helper_test.rb]: leftover?

**Comment by ShimShtein** [comment]
Done! Rebased and removed leftovers.

File: app/helpers/application_helper.rb [hunk 15]
```diff
@@ -441,10 +441,11 @@ def accessible_resource_records(resource, order = :name)
     klass.authorized.reorder(order)
   end
 
-  def accessible_resource(obj, resource, order = :name)
+  def accessible_resource(obj, resource, order = :name, opts = {})
```
Comment (tbrisker) [inline app/helpers/application_helper.rb]: you can use a keyward arg that defaults to resource here instead of a option hash

**Review by tbrisker — CHANGES_REQUESTED** [review]
Just one minor comment and we're good to go :+1:

**Comment by ShimShtein** [comment]
Done!

**Review by tbrisker — APPROVED** [review]

**Comment by tbrisker** [comment]
Thanks you very much @ShimShtein , sorry for the long review!

**Comment by dLobatog** [comment]
This broke the Host UI endpoints Katello needed to provision using synced content 😞 http://projects.theforeman.org/issues/18141
